doc/unboxing.md

```diff
@@ -0,0 +1,234 @@
+without causing bad behaviour (e.g. pushing allocations into loops).
+It currently supports:
+
+- Unboxing along non-looping control flow, within a single function, of the
+  following varieties of values:
+  - Boxed `float`
+  - Boxed `int32`
+  - Boxed `int64`
+  - Boxed `nativeint`
+  - Variants
+  - Closures (need examples).
+
+- Unboxing around loops, save that the corresponding allocations will only be
+  removed if:
+  - the first use of the value being unboxed in the loop can make use of the
+    unboxed form (for example if it is a floating-point arithmetic
+    operation, as opposed to something like a non-inlined function call); and
+  - this same condition also holds for the first use of the value after
+    the loop, on all exit paths from the loop.  (The trick of "`+. 0.`" can
+    be used here; this will not be optimised away by Flambda 2.)
+
+The unboxing code does not handle the untagging of integers.
+
+In the future support is expected for:
+
+- Unboxed returns from functions (e.g. returning values of variant type
+  without allocating).
+
+- Removal of allocations from loops where there does not always exist a use
+  of the corresponding unboxed value subsequent to the loop.  This
+  optimisation is expected to be based on partial dead code elimination.
+
+## Control over unboxing
+
+Unboxing depth.
+
+## Decision procedure
+
+The first step is computing the unboxing decision, and then in a second step,
+the decision is used to compute the extra params and args for the
+continuation. This is useful, in order to enable unboxing of recursive
+continuation. Indeed, in the case of a recursive continuation, we will have
+the following scenario: the first step (creating the unboxing decisions) is
+done before entering the handler of the continuation whose parameters we want
+to unbox, and thus we have access to all the non-recursive uses of the
+continuation. Then, once we have finished the downwards pass on the handler
+of the continuation, and thus have finally access to all uses of the
+continuation, we can compute the extra params and args for all use sites.
+
+The creation of unboxing decisions is done using the following steps:
+
+- Using the typing env computed from the continuation uses join, we first
+create an optimistic unboxing decision, based on the shape of the type of each
+of the continuation parameters.
+
+- We then do a pass on each decision which does two things:
+  - pre-compute the extra args for each of the continuation use-site
+  available at that point (i.e. non-recursive uses)
+  - extra_args can be either already in scope, or in some cases need to add
+  an extra primitive (i.e. untag, unbox, block_load, ...)
+
+- If it is not possible to generate such an extra arg (2), remove the
+problematic decision (and associated sub-decisions).
+
+- We then check that for each decision, at least one of the extra args
+computed is "already in scope", else the unboxing is not beneficial.
```

Comment:
In the case of loops, this check is skipped (because it is far more insteresting to keep the potential to unbox inside the loop, even if all [apply_cont]s outside the loop need to introduce a primitive at the call-sites.
-Then again, all the information about loops is subject to how #417  is eventually merged)